class-21.md

```diff
@@ -0,0 +1,44 @@
+* Edges (Directed or un-directed connections between those points)
+
+![DAG](https://cdn-images-1.medium.com/max/800/1*zzw_G4qNuVQaWXjXDYPjgg.png)
+
+We define a cycle in a graph if any vertex can be traced back to itself through a collection of edges, following their direction. A graph is said to be acyclic if no such combination of edges or verticies exists to fulfill this property. A very directed acyclic graph is a linked list, our traditional blockchain structure. Even if there are multiple chains branching from a point, cycles will not occur in this.
+
+![Blockchain](https://cdn-images-1.medium.com/max/800/1*srYXHCUC-FKvFcGCPMisrw.png)
+
+#### Taking advantage of a DAG
+
+We can use this structure to avoid validating all transactions in real time and waiting to accumulate them into blocks and distribute through the network. In IOTA, this is done by forcing each transaction to conform two previous transactions (detailed more [here](https://vxcompany.com/2018/05/23/iota-deepdive-1-consensus-on-the-tangle/)). In this structure, each transaction validates the existence of a previous transaction, and transactions afterwards further assert that the previous transaction exists. A transaction reaches a point of being fully consistent when it has enough references that have all validated it (both direct and indirect) to achieve a satisfactory degree of interaction for the application.
+
+In it's current state, IOTA has a set of training wheels on through the form of an orderer. While all transactions could theoretically be validated asynchronously through the DAG structure (called "the tangle" in IOTA), it makes use of a Zookeeper instance to force order/consistency on the network. To avoid a single transaction to be submitted once with a few references older in the chain, then again with many references right after (effectively trying to double spend), the network is using a ordering service commonly employed by distributed databases. There are plans, however, to remove this in the future.
+
+### Ripple
+
+In the Ripple blockchain, a set of transactions are all proposed for eventual candidacy in a block. Instead of taking turns signing off on blocks, all transactions are aggregated to a single pool, and those that receive 80% of votes are selected for a block. This system tolerates a 20% rate of bad actors occurring in the system and issuing byzantine faulty data.
+
+### Masternodes
+
+Dash and PivX don't necessarily have a distinctly different consensus mechanism from the others, however there are extra network effects in the form of masternodes. A masternode is often setup by paying in cryptocurrency and putting up a stake, and taking extra responsibilities in the network in accordance with that stake. The staked amount of tokens can be subject to forfeiture if the masternode fails to perform network duties in some instances. To compensate for this, the masternode is then eligible for some percentage of the block rewards.
```

Comment:
`Dash and PivX don't necessarily have a distinctly different consensus mechanism from the others, however there are extra network effects in the form of masternodes.`
-> Dash and PivX don't necessarily have a distinctly different consensus mechanism from the others. However, there are extra network effects in the form of masternodes.

Reply:
`A masternode is often setup by paying in cryptocurrency and putting up a stake, and taking extra responsibilities in the network in accordance with that stake.`
-> A masternode is often setup by paying cryptocurrency and putting up a stake in order to take extra responsibilities in the network in accordance with said stake.